### PR TITLE
ML DataLoaders

### PR DESCRIPTION
**Pull request by al-rigazzi**
This PR adds ML Data Loaders for Keras/TF and PyTorch.

They follow the download-and-store approach, not the streaming one (thus this could be added). Samples are downloaded from the DB and kept as copy of the data set (which means that they are not discarded after one epoch).

Specializations required some fine tuning, but the behavior is the same for TF and PyTorch. There are super-classes for all downloaders.

### CODE REVIEW

**Review by Spartee — COMMENTED** [review]
Will you run `make style` on this before I place in review comments?

**Comment by Spartee** [comment]
Went to review and looks like there are still a couple conflicts @al-rigazzi

**Review by Spartee — CHANGES_REQUESTED** [review]
Couple things

1. make sure to throw the deprecation wanring for ``smartsim.tf`` and lets let the user import that for a release. basically an empty module with the dep and an import works here. use the ``warnings`` library and throw ``DeprecationWarning``
2. Some re-naming for brevity
3. Lets add tests for all of these in the ``tests/backends`` folder.
4. Use the logger.
5. Lotta copy paste, lets keep the comments to what users need to see and direct them to the places they need to go for more information if necessary. These should be easy to grok in a VSCode like editor
6. instead of ``smartsim.ml.tf.data`` lets have those imports resolve to ``smartsim.ml.tf`` for brevity.
7. Workers comment may be in a differnt ticket but lets expose that if the functionality works and be sure to document the decision there.
8. tests tests tests tests

**Review by MattToast — COMMENTED** [review]
Supper cool looking stuff!! Here's some quick notes for consideration:

- I'm gonna second shortening up the docstrings to essential information. As someone coming from the outside, it took me while to try and figure out how the sub-classes differed while searching through the boilerplate.
- Found a couple of minor errors in the docs.
- Marked some places in where I thought the code-style seemed a bit off. Address as you see fit.

File: smartsim/ml/data.py [hunk 1]
```diff
@@ -0,0 +1,665 @@
...
+    """
+
+    def __init__(
+        self,
+        name="training_data",
+        sample_prefix="samples",
+        target_prefix="targets",
+        num_classes=None,
+        producer_prefixes=None,
+        smartredis_cluster=True,
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply smartsim/ml/data.py]: Done

File: smartsim/ml/data.py [hunk 2]
```diff
@@ -0,0 +1,665 @@
...
+                        then it is converted to `range(sub_indices)`.
+    :type sub_indices: int or list
+    :param num_classes: Number of classes of targets, if categorical
+    :type num_classes: int
+    :param producer_prefixes: Prefixes of processes which will be producing batches.
+                              This can be useful in case the consumer processes also
+                              have other incoming entities.
+    :type producer_prefixes: str
+    :param smartredis_cluster: Whether the Orchestrator will be run as a cluster
+    :type smartredis_cluster: bool
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply smartsim/ml/data.py]: Done

File: smartsim/ml/data.py [hunk 3]
```diff
@@ -0,0 +1,665 @@
...
+    :param smartredis_address: Address of Redis client as <ip_address>:<port>
+    :type smartredis_address: str
+    :param replica_rank: When BatchDownloader is used in a distributed setting, indicates
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply smartsim/ml/data.py:617]: Yeah, this was confusing. The sub-indices are those of the uploader, I renamed them as `uploader_ranks`, which is just an `int`.

File: smartsim/ml/torch/data.py [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+    def __init__(self, dataset: StaticDataGenerator, **kwargs):
+        super().__init__(
+            dataset,
+            worker_init_fn=self.worker_init_fn,
+            persistent_workers=True,
+            **kwargs,
+        )
+
+    @staticmethod
+    def worker_init_fn(worker_id):
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply smartsim/ml/torch/data.py:247]: Both are handled by the super-class `torch.DataLoader` and corresponding settings.

File: tutorials/04_ml_training/tf/data_uploader.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
+from smartsim.ml import TrainingDataUploader
+from os import environ
+from time import sleep
+import numpy as np
+from mpi4py import MPI
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply tutorials/04_ml_training/tf/data_uploader.py:5]: I put it in the Jupyter notebook, should be enough, but I'm open to suggestions.

File: smartsim/ml/data.py [hunk 6]
```diff
@@ -0,0 +1,665 @@
+import time
+from os import environ
+
+import numpy as np
+
+from smartredis import Client, Dataset
+from smartredis.error import RedisReplyError
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply smartsim/ml/data.py]: Removed

File: tutorials/04_ml_training/torch/training_service_hvd.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
+import numpy as np
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply tutorials/04_ml_training/torch/training_service_hvd.py]: Fixed

File: tutorials/04_ml_training/torch/training_service.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
+import numpy as np
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply tutorials/04_ml_training/torch/training_service.py]: Fixed

File: tutorials/04_ml_training/tf/training_service_hvd.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
+import numpy as np
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply tutorials/04_ml_training/tf/training_service_hvd.py]: Fixed

File: smartsim/ml/torch/data.py [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+                ]
+            else:
+                sources = overall_sources[worker_id * per_worker :]
+        else:
+            if worker_id < len(overall_sources):
+                sources = overall_sources[worker_id]
+            else:
+                sources = []
+
+        print(f"{worker_id}: {sources}")
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply smartsim/ml/torch/data.py]: Well, this was too verbose anyhow. I removed it.

File: doc/tutorials/machine_learning.rst [hunk 11]
```diff
@@ -351,3 +351,160 @@ RandomForestRegressor. As with the other examples, the skl2onnx function
...
+within a SmartSim ``Experiment``, we can use the following code:
+
+.. code-block:: python
+    
+    from smartsim import Experiment
+    from smartsim.database import Orchestrator
+    from smartsim.settings import RunSettings
+
+    db = Orchestrator(port=6780)
+    exp = ("online-training", launcher="local")
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply doc/tutorials/machine_learning.rst:481]: Good catch

File: doc/tutorials/machine_learning.rst [hunk 12]
```diff
@@ -351,3 +351,160 @@ RandomForestRegressor. As with the other examples, the skl2onnx function
...
+
+    # producer
+    producer_script = "producer.py"
+    settings = RunSettings("python", exe_args=producer_script)
+    uploader_model = exp.create_model("producer", settings)
+    uploader_model.attach_generator_files(to_copy=script)
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply doc/tutorials/machine_learning.rst:487]: Fixed

File: doc/tutorials/machine_learning.rst [hunk 13]
```diff
@@ -351,3 +351,160 @@ RandomForestRegressor. As with the other examples, the skl2onnx function
...
+    uploader_model.enable_key_prefixing()
+
+    # training_service
+    training_script = "training_service.py"
+    settings = RunSettings("python", exe_args=training_script)
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply doc/tutorials/machine_learning.rst:492]: Done

File: smartsim/ml/data.py [hunk 14]
```diff
@@ -0,0 +1,665 @@
...
+    def __iter__(self):
+
+        if not self.sources:
+            pass
+        else:
+            self.update_data()
+            # Generate data
+            if len(self) < 1:
+                raise ValueError(
+                    "Not enough samples in generator for one batch. Please run init_samples() or initialize generator with init_samples=True"
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply smartsim/ml/data.py:361]: Yep, split over two lines, thanks

File: smartsim/ml/data.py [hunk 15]
```diff
@@ -0,0 +1,665 @@
...
+    def __len__(self):
+        length = int(np.floor(self.num_samples / self.batch_size))
+        return length
+
+
+    def __iter__(self):
+
+        if not self.sources:
+            pass
+        else:
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply smartsim/ml/data.py:356]: Removed

File: smartsim/ml/tf/data.py [hunk 16]
```diff
@@ -0,0 +1,248 @@
+import time
+from os import environ
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply smartsim/ml/tf/data.py]: Fixed

File: smartsim/ml/torch/data.py [hunk 17]
```diff
@@ -0,0 +1,279 @@
...
+        )
+
+    @staticmethod
+    def worker_init_fn(worker_id):
+        worker_info = torch.utils.data.get_worker_info()
+        dataset = worker_info.dataset  # the dataset copy in this worker process
+        dataset.init_sources()
+        overall_sources = dataset.sources
+
+        worker_id = worker_info.id
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply smartsim/ml/torch/data.py:253]: I defer to the Torch docs I took this from. My best guess is that it is not.

File: doc/tutorials/machine_learning.rst [hunk 18]
```diff
@@ -351,3 +351,160 @@ RandomForestRegressor. As with the other examples, the skl2onnx function
...
+    from smartsim import Experiment
+    from smartsim.database import Orchestrator
+    from smartsim.settings import RunSettings
+
+    db = Orchestrator(port=6780)
+    exp = ("online-training", launcher="local")
+
+    # producer
+    producer_script = "producer.py"
+    settings = RunSettings("python", exe_args=producer_script)
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply doc/tutorials/machine_learning.rst]: Fixed

File: doc/tutorials/machine_learning.rst [hunk 19]
```diff
@@ -351,3 +351,160 @@ RandomForestRegressor. As with the other examples, the skl2onnx function
...
+    uploader_model = exp.create_model("producer", settings)
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply doc/tutorials/machine_learning.rst]: I had no clue we could do that.

File: doc/tutorials/machine_learning.rst [hunk 20]
```diff
@@ -351,3 +351,160 @@ RandomForestRegressor. As with the other examples, the skl2onnx function
...
+    uploader_model.enable_key_prefixing()
+
+    # training_service
+    training_script = "training_service.py"
+    settings = RunSettings("python", exe_args=training_script)
+    trainer_model = exp.create_model("training_service", settings)
+    trainer_model.register_incoming_entity(uploader_model)
+
+    exp.start(db)
+    exp.start(uploader_model, block=False, summary=False)
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply doc/tutorials/machine_learning.rst]: Done

File: smartsim/ml/tf/data.py [hunk 21]
```diff
@@ -0,0 +1,248 @@
...
+    :param init_samples: whether samples should be initialized in the constructor
+    :type init_samples: bool
+    """
+
+    def __init__(self, **kwargs):
+        StaticDataGenerator.__init__(self, **kwargs)
+
+
+    def __data_generation(self, indices):
+        return StaticDataGenerator.__data_generation(self, indices)
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply smartsim/ml/tf/data.py]: Very good catch, it was legacy code

File: smartsim/ml/data.py [hunk 22]
```diff
@@ -0,0 +1,665 @@
...
+    By default, the ContinuousBatchDownloader has to be created in a process
+    launched through SmartSim, with sample producers listed as incoming
+    entities.
+
+    All details about the batches must be defined in
+    the constructor; two mechanisms are available, `manual` and
+    `auto`.
+
+     - When specifying `auto`, the user must also specify
+      `uploader_name`. BatchDownloader will get all needed information
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply smartsim/ml/data.py]: Yep, now I fixed all names.

File: smartsim/ml/data.py [hunk 23]
```diff
@@ -0,0 +1,665 @@
...
+        else:
+            raise ValueError(
+                f"uploader_info must be one of 'auto' or 'manual', but was {self.uploader_info}"
+            )
+
+        if init_samples:
+            self.init_sources()
+            self.init_samples()
+
+    def log(self, message):
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply smartsim/ml/data.py:273]: OK, added logger now!

File: doc/tutorials/machine_learning.rst (hunk 20 above)
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply doc/tutorials/machine_learning.rst]: Done

**Review by Spartee — APPROVED** [review]
LGTM! Great work! Excited to see the propagate into the community.

**Pull request merged**